src/lib/core/util.c

```diff
@@ -484,3 +484,42 @@ thread_sleep(double sec)
 	assert(rc == 0);
 	(void)rc;
 }
+
+/*
+ * 32 4 kilobyte pages seems to be the limit for a single environment variable
+ * or argv element set by the kernel. We use that to limit maximum buffer size
+ * allocated for an env variable.
+ */
+#define MAX_ENV_VAR_SIZE 131072
+
+char *
+getenv_safe(const char *name, char *buf, size_t buf_size)
+{
+	assert(buf != NULL || buf_size == 0);
+	assert(name != NULL);
+
+	char *var = getenv(name);
+	if (var == NULL)
+		return NULL;
+	if (buf == NULL)
+		buf_size = MAX_ENV_VAR_SIZE;
+	size_t var_len = strnlen(var, buf_size);
+	if (var_len >= buf_size) {
+		say_warn("Ignoring environment variable %s because its value "
+			 "is too long (>= %zu)", name, buf_size);
+		return NULL;
+	}
+	char *alloc_buf = NULL;
+	if (buf == NULL) {
+		alloc_buf = xmalloc(var_len + 1);
+		buf = alloc_buf;
+	}
+	memcpy(buf, var, var_len + 1);
+	if (buf[var_len] != '\0') {
+		say_warn("Ignoring environment variable %s because it changed "
+			 "length during copying (expected %zu)", name, var_len);
+		free(alloc_buf);
+		return NULL;
```

Comment:
I don't like this branch, because it isn't tested. Maybe, just set `buf[var_len]` to 0 silently?

Reply:
Ok, my first thoughts were that returning nothing is better than returning a possibly truncated value. But then I understood that the value might be malformed anyway.
So this seems to be better indeed:
```diff
diff --git a/src/lib/core/util.c b/src/lib/core/util.c
index a3067cd65..916e3d737 100644
--- a/src/lib/core/util.c
+++ b/src/lib/core/util.c
@@ -516,10 +516,8 @@ getenv_safe(const char *name, char *buf, size_t buf_size)
        }
        memcpy(buf, var, var_len + 1);
        if (buf[var_len] != '\0') {
-               say_warn("Ignoring environment variable %s because it changed "
-                        "length during copying (expected %zu)", name, var_len);
-               free(alloc_buf);
-               return NULL;
+               /* The value was changed while copying.*/
+               buf[var_len] = '\0';
        }
        return buf;
 }
```

Reply:
> I don't like this branch, because it isn't tested. Maybe, just set `buf[var_len]` to 0 silently?

Disregarding of behavior we'll choose, this situation will not become tested until we'll test it. (But, well, we can write the code without branching here and the `buf[var_len] = '\0';` line will technically be covered -- but actually only when it already contains `'\0'`.)

> Ok, my first thoughts were that returning nothing is better than returning a possibly truncated value.

Seems perfectly valid.

> But then I understood that the value might be malformed anyway.

The value may be changed and a caller will get a result just like it would call the function a moment later if the lengths are the same. Otherwise it'll get `NULL` (plus the warning) and it seems better than truncation.

I like the previous solution more and it is four lines of code -- I hope it does not deserve a special effort to test it.

Reply:
> > But then I understood that the value might be malformed anyway.
> 
> The value may be changed and a caller will get a result just like it would call the function a moment later if the lengths are the same.

AFAIU theoretically `memcpy` may race with `setenv` in such a way that the user will get half of the old and half of the new value in the buffer, i.e. the check for the length isn't enough to guarantee consistency.

> ```diff
> diff --git a/src/lib/core/util.c b/src/lib/core/util.c
> index a3067cd65..916e3d737 100644
> --- a/src/lib/core/util.c
> +++ b/src/lib/core/util.c
> @@ -516,10 +516,8 @@ getenv_safe(const char *name, char *buf, size_t buf_size)
>         }
>         memcpy(buf, var, var_len + 1);
>         if (buf[var_len] != '\0') {
> -               say_warn("Ignoring environment variable %s because it changed "
> -                        "length during copying (expected %zu)", name, var_len);
> -               free(alloc_buf);
> -               return NULL;
> +               /* The value was changed while copying.*/
> +               buf[var_len] = '\0';
>         }
>         return buf;
>  }
> ```

Nit: The `if` isn't necessary. Also, I'd expand on the comment.

```C
memcpy(buf, var, var_len);
/*
 * The value could have changed during copying so we reset the terminating
 * nul explicitly to avoid out-of-bounds access. The returned value may still
 * be malformed, but there's nothing we can really do about it.
 */
buf[var_len] = 0;
```

Reply:
> Disregarding of behavior we'll choose, this situation will not become tested until we'll test it. 

@Totktonada, this is true, but still I agree with @locker here.
Even if we return `NULL` when we detect line length inconsistency, we don't guard the user from truncated values: an attacker or an erroneous script may paste a truncated value to env anyway. So the user of `getenv` or `getenv_safe` should be ready for malformed values.

Another point: returning `NULL` when we detected changed line length looks like a promise to the user, that when a value is returned, it's valid. This is not true. As @locker mentioned above, line length might stay the same, but the value will contain garbage.

We could add some more validation, like this:

```c
buf[var_len] = '\0';
if (strcmp(buf, var) != 0) {
    say_warn("Ignoring environment variable, because it was changed while copying.");
    return NULL;
}
```

This means that the value we returned is the same as the one contained in environment. But still, seems there is a tiny chance that `setenv` races with `stcmp`. So I'm not sure this is worth implementing.

And even with such checks the user has to validate the returned string later on. We only guard the user from accidental string corruption, but not from attackers, who can corrupt the string intentionally.

Hence the latest diff:
```diff
diff --git a/src/lib/core/util.c b/src/lib/core/util.c
index 916e3d737..f6db14219 100644
--- a/src/lib/core/util.c
+++ b/src/lib/core/util.c
@@ -515,9 +515,13 @@ getenv_safe(const char *name, char *buf, size_t buf_size)
                buf = alloc_buf;
        }
        memcpy(buf, var, var_len + 1);
-       if (buf[var_len] != '\0') {
-               /* The value was changed while copying.*/
-               buf[var_len] = '\0';
-       }
+       /*
+        * The value could have changed during copying so we reset the
+        * terminating zero explicitly to avoid out-of-bounds access.
+        * The returned value may still be malformed, intentionally, for
+        * example, but there's nothing we can really do about it. The user must
+        * check that the returned value is valid for his use.
+        */
+       buf[var_len] = '\0';
        return buf;
 }
```

Reply:
Seems so, we should do `strcmp()`/`memcmp()` to actually guarantee that the copy is consistent. I see no problem in `strcmp()` and `setenv()` race: I guess it only means that we can say that a valid copy is invalid, not vice versa.

In my understanding 'getenv_safe()' is something really safe, but I'm okay to leave it as is for now if actual guarantees will be clarified in the function description: what 'safe' means and what it does not mean.

Reply:
> In my understanding 'getenv_safe()' is something really safe, but I'm okay to leave it as is for now if actual guarantees will be clarified in the function description: what 'safe' means and what it does not mean.

My point still stands: we could only guarantee that a copy is valid, but the user still has to check that the actual value is valid as well. So IMO there's no point in guaranteeing a valid copy of an possibly invalid value (by invalid I mean crafted by an attacker or simply erroneous).

Reply:
After a verbal discussion with @Totktonada decided to stick with the current approach (truncation) and add a comment:

```diff
diff --git a/src/trivia/util.h b/src/trivia/util.h
index 614e52805..629f4403c 100644
--- a/src/trivia/util.h
+++ b/src/trivia/util.h
@@ -655,6 +655,16 @@ thread_sleep(double sec);
  *         NULL in case no value is found.
  *              in case buf != NULL and strlen(value) >= \a buf_size.
  *              in case buf == NULL and strlen(value) >= internal limit.
+ *
+ * When a non-null pointer is returned, it's guaranteed to contain a
+ * null-terminated string. The string is a copy of the corresponding environment
+ * variable in all cases, except when `getenv_safe` is run concurrently with
+ * `setenv`.
+ * In that case the buffer might contain:
+ * - an old variable value,
+ * - a new value, truncated to not exceed old value length,
+ * - garbage, truncated to not exceed old value length
+ * Hence the user has to validate the returns.
  */
 char *
 getenv_safe(const char *name, char *buf, size_t buf_size);
```